code/modules/power/supermatter/supermatter.dm

```diff
@@ -200,102 +197,204 @@ GLOBAL_DATUM(main_supermatter_engine, /obj/machinery/power/supermatter_crystal)
 
 	soundloop = new(src, TRUE)
-	if(ispath(psyOverlay))
-		psyOverlay = new psyOverlay()
-	else
-		stack_trace("Supermatter created with non-path psyOverlay variable. This can break things, please fix.")
-		psyOverlay = new()
 
 	if (!moveable)
 		move_resist = MOVE_FORCE_OVERPOWERING // Avoid being moved by statues or other memes
 
-	update_constants()
+	// Damn math nerds
+	powerloss_linear_threshold = sqrt(POWERLOSS_LINEAR_RATE / 3 * POWERLOSS_CUBIC_DIVISOR ** 3)
+	powerloss_linear_offset = -1 * powerloss_linear_threshold * POWERLOSS_LINEAR_RATE + (powerloss_linear_threshold / POWERLOSS_CUBIC_DIVISOR) ** 3
 
 /obj/machinery/power/supermatter_crystal/Destroy()
 	if(warp)
 		vis_contents -= warp
 		QDEL_NULL(warp)
 	investigate_log("has been destroyed.", INVESTIGATE_ENGINE)
 	SSair.stop_processing_machine(src)
+	absorbed_gasmix = null
 	QDEL_NULL(radio)
 	QDEL_NULL(countdown)
 	if(is_main_engine && GLOB.main_supermatter_engine == src)
 		GLOB.main_supermatter_engine = null
 	QDEL_NULL(soundloop)
-	if(psyOverlay)
-		QDEL_NULL(psyOverlay)
 	return ..()
 
-/obj/machinery/power/supermatter_crystal/proc/update_constants()
-	pressure_bonus_derived_steepness = (1 - 1 / pressure_bonus_max_multiplier) / (pressure_bonus_max_pressure ** pressure_bonus_curve_angle)
-	pressure_bonus_derived_constant = 1 / pressure_bonus_max_multiplier - pressure_bonus_derived_steepness
-	powerloss_linear_threshold = sqrt(POWERLOSS_LINEAR_RATE / 3 * POWERLOSS_CUBIC_DIVISOR ** 3)
-	powerloss_linear_offset = -1 * powerloss_linear_threshold * POWERLOSS_LINEAR_RATE + (powerloss_linear_threshold / POWERLOSS_CUBIC_DIVISOR) ** 3
-
 /obj/machinery/power/supermatter_crystal/examine(mob/user)
 	. = ..()
 	var/immune = HAS_TRAIT(user, TRAIT_MADNESS_IMMUNE) || (user.mind && HAS_TRAIT(user.mind, TRAIT_MADNESS_IMMUNE))
-	if(isliving(user) && !immune && (get_dist(user, src) < HALLUCINATION_RANGE(power)))
+	if(isliving(user) && !immune && (get_dist(user, src) < SM_HALLUCINATION_RANGE(internal_energy)))
 		. += span_danger("You get headaches just from looking at it.")
 	. += delamination_strategy.examine(src)
 	return .
 
+/obj/machinery/power/supermatter_crystal/process_atmos()
+	// PART 1: PRELIMINARIES
+	if(disable_process)
+		return
+
+	var/turf/local_turf = loc
+	if(!istype(local_turf))//We are in a crate or somewhere that isn't turf, if we return to turf resume processing but for now.
+		return  //Yeah just stop.
+	if(isclosedturf(local_turf))
+		var/turf/did_it_melt = local_turf.Melt()
+		if(!isclosedturf(did_it_melt)) //In case some joker finds way to place these on indestructible walls
+			visible_message(span_warning("[src] melts through [local_turf]!"))
+		return
+
+	// PART 2: GAS PROCESSING
+	var/datum/gas_mixture/env = local_turf.return_air()
+	absorbed_gasmix = env?.remove_ratio(absorption_ratio) || new()
+	absorbed_gasmix.volume = (env?.volume || CELL_VOLUME) * absorption_ratio // To match the pressure.
```

Comment:
this really feels like an if(env) moment

Reply:
Also why the || CELL_VOLUME? what case are you trying to catch there?

Reply:
null || 2500 or 0 || 2500, just in case.

Reply:
how would 0 get in there?

Reply:
Just in case to be honest, mainly thinking about the null thing rather than zero when i was adding it, didnt like using ifs so much